File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        water = 0
        left_high = right_high = 0
        left, right = 0, len(height) - 1
        while left < right:
            if height[left] < height[right]:
                if height[left] < left_high:
                    water += left_high - height[left]
                else:
                    left_high = height[left]
                left += 1
            else:
                if height[right] < right_high:
                    water += right_high - height[right]
                else:
                    right_high = height[right]
                right -= 1
        return water

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int result = 0;
        int left = 0;
        int right = height.size() - 1;
        int max_left_h = 0;
        int max_right_h = 0;

        while (left < right) {
            if (height[left] < height[right]) {
                max_left_h = max(max_left_h, height[left]);
                result += max_left_h - height[left];
                left += 1;
            } else {
                max_right_h = max(max_right_h, height[right]);
                result += max_right_h - height[right];
                right -= 1;
            }
        }
        return result;
    }
};
